### DISCUSSION TITLE
RP2040 MP3 WAV noisy or breaking up?

### DISCUSSION THREAD

**Discussion by PJ5168** [earlephilhower/ESP8266Audio#603] [General]
Dear Folks,

ESP32 and ESP32 S3 are doing great playing the WAV and MP3 files using this library.

But RP2040 133Mhz is noisy? Seems that the MCU speed is not fast enough?
Then, run the RP2040 at 240Mhz, the sound is breaking up/pop? The SD is running at 40Mhz, reading around 4MB/s.
Are there something wrong with the PIO I2S?

Best,

PJ.

**Comment by knuf-kandler** [comment]
Similar problems here. 

I2S example from pico-core (generating 440Hz square) - OK
MOD to I2s from memory (esp8266 example) - OK with some clicks/pops now and then
WAV from memory to I2s - almost OK with few clicks/pops
MIDI from LittleFS - terrible robotic sounding reproduction
MP3 from LittleFS - robotic reproduction - overclocked or not-overclocked
WAV from LittleFS - seems like WAVs at 22kHz mono are playing better than 44kHz stereo - both unusuable - noisy and crackling

tried with two max98357A and two breadboards, powering it from 15W source

**Reply by earlephilhower** [reply]
Try adding `I2S.setBuffers(5, 64);` at the beginning of `setup()` to increase the size of individual I2S buffers.  Clicking means the I2S is underflowing and, assuming the Pico@133MHz is fast enough, making each buffer larger should help.

**Reply by PJ5168** [reply]
Dear Sir,

Thanks for the prompt reply. Appreciate!
Use following settings :
I2S i2s(OUTPUT);
i2s.setBuffers(5, 64); 

Still, the same, even set buffers to 8, 128 or 16, 256.
The 133Mhz seems not fast enough, still noisy and crackling.
The 240Mhz, not noisy but same crackling.

Or using following : same crackling output at 240Mhz
buff = new AudioFileSourceBuffer(source, 8192);    // 2 4 8 16 32 64K, same on MP3 and WAV....
decoder->begin(buff, output);                                //

Thanks again,

PJ.

**Reply by knuf-kandler** [reply]
Thanks for help!

Same here. Tried different settings/combinations for buffers settings (FileSource vs i2s) with no luck.

at 120MHz - all WAVs (22kHz, 44kHz, 16b / 8b, mono/stereo) produce noisy and hissing output
at 240Mhz - sound is less noisy/hissy but very fast periodic crackling is more pronounced 

strange things
 - changes between lsb and msb modes do nothing - it sounds the same
 - feeding it with zero samples WAV (~15s long) produces silence with no crackles, noise or hiss.

**Reply by knuf-kandler** [reply]
some more experiments. to be able to continue with my project i tried pwm output on pico using examples from esp82266audio and also from rp2040 core (by the same author of course (: ) and found that playback is very clean 

but - only from progmem (tried MOD and WAV playback)

when i try to load files from LittleFS i get exactly the same results as with MAX98357 : 

@120 MHz - playback suffers from terrible noise and hiss 
@ 240 MHz - no hiss or noise, but very fast periodic crackling

**Reply by PJ5168** [reply]
Same here. The 240Mhz RP2040 is fast enough to read from SD, decode the audio file, and feed the buffers.
Sounds like the RP2040 I2S PIO has timing issue!?
Just like ESP32, there is no way to set a close enough clock for 192K I2S.

**Comment by MrTucks** [comment]
I think I may have the same problem.
First I tried to play a mp3 from an SD-Card and I got no music at all, just some cracking that started a few seconds after starting playback. (There was a bit of wifi stuff in my code as well, so maybe there was just not enough processing power left.)

To rule that (and the SD card) out, I tried to get the [PlayWAVFromPPROGMEM example](https://github.com/earlephilhower/ESP8266Audio/tree/master/examples/PlayWAVFromPROGMEM) working, except with an i2s output instead of i2sNoDac. It was way too loud, so I added a mixer and set the gain to 0.1.

There was a lot of cracking. 
Adding `I2S.setBuffers(5, 64);` in the beginning of `setup()` in AudioOutputI2S.cpp didn't really help.
`i2s.setBuffers(5,1024);`: noticeably decreased cracking, however it didn't get it to zero. 
`i2s.setBuffers(5,2048);`: introduced a kind of echo at the end of the track. (And at the beginning of the track, it's causing a small pop or missing a few samples or something like that.)

How I tested that my Hardware works flawlessly:
[Here, from Waveshare](https://www.waveshare.com/wiki/Pico-Audio) I downloaded a .uf2 that turns the Pico into a USB sound card.
With that uf2, it played back audio from my PC without any issues and zero cracking/popping. 
Sadly, there seems to be no source code (or I didn't find it). I guess it's using [the .pio from pico-extras.](https://github.com/raspberrypi/pico-extras/blob/master/src/rp2_common/pico_audio_i2s/audio_i2s.pio) And I guess maybe it's doing less processing?

Maybe the Pico is just not fast enough for that stuff?
I didn't try any overclocking yet.

P.S: This is my first time using the Pico or i2s. Sorry if I'm missing something.
Software:
https://github.com/earlephilhower/arduino-pico/ with platformio like described [in the docs.](https://arduino-pico.readthedocs.io/en/latest/platformio.html)
Hardware:
Pi Pico W with one [Adafruit MAX98357 I2S Class-D Mono Amp](https://learn.adafruit.com/adafruit-max98357-i2s-class-d-mono-amp)

My code:
```cpp
#include <Arduino.h>

#include <AudioFileSourcePROGMEM.h>
#include <AudioGeneratorWAV.h>
#include <AudioOutputI2S.h>
#include <AudioOutputSerialWAV.h>
#include <AudioOutputMixer.h>

// VIOLA sample taken from https://ccrma.stanford.edu/~jos/pasp/Sound_Examples.html
#include <viola.h>

AudioGeneratorWAV *wav;
AudioFileSourcePROGMEM *file;
AudioOutputMixer *mixer;
AudioOutputMixerStub *stub;
AudioOutputI2S *i2s;

void setup()
{
  Serial.begin(115200);
  delay(1000);
  Serial.printf("WAV start\n");

  //audioLogger = &Serial;
  file = new AudioFileSourcePROGMEM( viola, sizeof(viola) );
  i2s = new AudioOutputI2S(8000);
  
  wav = new AudioGeneratorWAV();
  mixer = new AudioOutputMixer(16,i2s);
  stub = mixer->NewInput();
  stub->SetGain(0.1);
  wav->begin(file, stub);
}

void loop()
{
  if (wav->isRunning()) {
    if (!wav->loop()) wav->stop();
  } else {
    Serial.printf("WAV done\n");
    delay(1000);
  }
}
```

**Reply by PJ5168** [reply]
Dear Sir,

I saw this one as well. But I did not give it a try. Thanks for letting me know the result.
Believe, you download the UF2 from the waveshare web page. The demo code is in 7Z file.
There is audio_pio.pio file in it. Also, there are other related file in the LIB folder.
2 cents,

PJ.

**Reply by JaredTS486** [reply]
> I think I may have the same problem. First I tried to play a mp3 from an SD-Card and I got no music at all, just some cracking that started a few seconds after starting playback. (There was a bit of wifi stuff in my code as well, so maybe there was just not enough processing power left.)
> 
> To rule that (and the SD card) out, I tried to get the [PlayWAVFromPPROGMEM example](https://github.com/earlephilhower/ESP8266Audio/tree/master/examples/PlayWAVFromPROGMEM) working, except with an i2s output instead of i2sNoDac. It was way too loud, so I added a mixer and set the gain to 0.1.
> 
> There was a lot of cracking. Adding `I2S.setBuffers(5, 64);` in the beginning of `setup()` in AudioOutputI2S.cpp didn't really help. `i2s.setBuffers(5,1024);`: noticeably decreased cracking, however it didn't get it to zero. `i2s.setBuffers(5,2048);`: introduced a kind of echo at the end of the track. (And at the beginning of the track, it's causing a small pop or missing a few samples or something like that.)
> 
> How I tested that my Hardware works flawlessly: [Here, from Waveshare](https://www.waveshare.com/wiki/Pico-Audio) I downloaded a .uf2 that turns the Pico into a USB sound card. With that uf2, it played back audio from my PC without any issues and zero cracking/popping. Sadly, there seems to be no source code (or I didn't find it). I guess it's using [the .pio from pico-extras.](https://github.com/raspberrypi/pico-extras/blob/master/src/rp2_common/pico_audio_i2s/audio_i2s.pio) And I guess maybe it's doing less processing?
> 
> Maybe the Pico is just not fast enough for that stuff? I didn't try any overclocking yet.
> 
> P.S: This is my first time using the Pico or i2s. Sorry if I'm missing something. Software: https://github.com/earlephilhower/arduino-pico/ with platformio like described [in the docs.](https://arduino-pico.readthedocs.io/en/latest/platformio.html) Hardware: Pi Pico W with one [Adafruit MAX98357 I2S Class-D Mono Amp](https://learn.adafruit.com/adafruit-max98357-i2s-class-d-mono-amp)
> 
> My code:
> 
> ```c++
> #include <Arduino.h>
> 
> #include <AudioFileSourcePROGMEM.h>
> #include <AudioGeneratorWAV.h>
> #include <AudioOutputI2S.h>
> #include <AudioOutputSerialWAV.h>
> #include <AudioOutputMixer.h>
> 
> // VIOLA sample taken from https://ccrma.stanford.edu/~jos/pasp/Sound_Examples.html
> #include <viola.h>
> 
> AudioGeneratorWAV *wav;
> AudioFileSourcePROGMEM *file;
> AudioOutputMixer *mixer;
> AudioOutputMixerStub *stub;
> AudioOutputI2S *i2s;
> 
> void setup()
> {
>   Serial.begin(115200);
>   delay(1000);
>   Serial.printf("WAV start\n");
> 
>   //audioLogger = &Serial;
>   file = new AudioFileSourcePROGMEM( viola, sizeof(viola) );
>   i2s = new AudioOutputI2S(8000);
>   
>   wav = new AudioGeneratorWAV();
>   mixer = new AudioOutputMixer(16,i2s);
>   stub = mixer->NewInput();
>   stub->SetGain(0.1);
>   wav->begin(file, stub);
> }
> 
> void loop()
> {
>   if (wav->isRunning()) {
>     if (!wav->loop()) wav->stop();
>   } else {
>     Serial.printf("WAV done\n");
>     delay(1000);
>   }
> }
> ```

What IDE did you use to compile this for the Pico? I am struggling with Platform.IO at the moment, it seems to think that the Pico is incompatible with the https://github.com/esp8266/Arduino framework.

**Reply by MrTucks** [reply]
The Pico uses the RP2040 processor from raspberry pi and not the ESP8266, so it needs a framework made for the RP2040.

I used Platformio but with the earlephilhower arduino-pico core.
Here are the instructions to set it up: [https://registry.platformio.org/platforms/platformio/raspberrypi](https://arduino-pico.readthedocs.io/en/latest/platformio.html)

The default Platformio Platform for the pico ([https://registry.platformio.org/platforms/platformio/raspberrypi](https://registry.platformio.org/platforms/platformio/raspberrypi))  is honestly very lacking. It barely supports the Pico W (can't even blink the onboard LED). And doesn't have I2S support (without additional libraries).

P.S: I didn't manage to do any more testing since my last post. Please write, if you manage to figure anything out.

**Reply by JaredTS486** [reply]
Thank you! I figured I was not supposed to use the ESP8266 framework but the library was not working with the default platform/board in PlatformIO.

**Comment by LinusHeu** [comment]
Could it be that you experienced this bug in arduino-pico which caused clicking/cracks?
https://github.com/earlephilhower/arduino-pico/pull/1500
https://github.com/earlephilhower/arduino-pico/discussions/1491

It's fixed in the [latest release](https://github.com/earlephilhower/arduino-pico/releases/tag/3.2.2).

**Reply by PJ5168** [reply]
Did you give it a try?
For 2.7.3, all the .ino files compiled OK.
But 3.2.2 with following error message??

Fatal Python error: initfsencoding: unable to load the file system codec
ModuleNotFoundError: No module named 'encodings'

**Reply by earlephilhower** [reply]
No, that's an IDE 1.8 issue and not related to anything here or in the core.  See https://github.com/earlephilhower/arduino-pico/issues/976#issuecomment-1312727031 for the steps to re-install the core properly.

**Reply by PJ5168** [reply]
Dear Sir,

Thanks for the prompt reply.
Followed all the steps accordingly but still got the same error message...
Arduino IDE : 1.8.15
OS : Windows 11 Pro, 22H2, build 22621.1635

Best Regards,

PJ.

**Reply by PJ5168** [reply]
The verbose compile output : 

"C:\\Users\\W480N2\\AppData\\Local\\Arduino15\\packages\\rp2040\\tools\\pqt-python3\\1.0.1-base-3a57aed/python3" -I "C:\\Users\\W480N2\\AppData\\Local\\Arduino15\\packages\\rp2040\\hardware\\rp2040\\3.2.2/tools/signing.py" --mode header --publickey "C:\\M5\\Arduino\\CoreMark/public.key" --out "C:\\Users\\W480N2\\AppData\\Local\\Temp\\arduino_build_413207/core/Updater_Signing.h"
Fatal Python error: initfsencoding: unable to load the file system codec
ModuleNotFoundError: No module named 'encodings'

**Reply by PJ5168** [reply]
Dear Sir,

Seems that the installation process does not correctly install all the python related files.
After manually re-install the python described in the #976 comment, now the error message is gone.

Yours,

PJ.

**Comment by PJ5168** [comment]
Dear Folks,

Please update to core 3.2.2.
If there is error message when compiling, please manually re-install the python in the tool directory.
Now, it's fine to playback the MP3 files with PCM5102A on 240Mhz Pico and 40Mhz SPI SD card.

Yours,

PJ.

**Reply by madias123** [reply]
Hello, 
what Bitrate could we expect with the RP2040 while playing MP3s? I need at least 192kBit/s to get the VS1053 out of my project ;) 
Thank you regards Matthias

@PJ5168 May you share you working example code? Thank you

**Reply by PJ5168** [reply]
Dear Madias123,

Not sure what you mean.
1. 192KBit/s or 192Ks/s?
2. 192KBit/s = 6Ks/s dual 16-bit channel
3. If you mean 192Ks/s, why use MP3 format? The MP3 format will lost most 192Ks/s information. Do you mean FLAC or WAV?
4. ESP32 APLL can support 96Ks/s 24-bit dual-channel FLAC which is 4.608Mbits/s.
5. For FLAC and WAV file, the size is really big. It takes time to read it in from slow SD card.
6. The 133Mhz Pico is slower than ESP32 S3. You may consider other faster MCU with audio hardware.
2 cents!~

PJ

**Reply by madias123** [reply]
I thought 192 kbps MP3, "standard" quality MP3 fixed bitrate.
Neverless, I'm fighting with     
`bool SetLsbJustified(bool lsbJustified);  // Allow supporting non-I2S chips, e.g. PT8211 `
in this library.
While setting this in the main pico core to:
`i2s.setLSBJFormat();`
I cannot get it to run with ESP8266Audio (always distorted Audio)
Things like this wont work:
```
 in = new AudioFileSourcePROGMEM(sampleaac, sizeof(sampleaac));
  aac = new AudioGeneratorAAC();
  out = new AudioOutputI2S();
out->SetLsbJustified(1); 
  aac->begin(in, out);
; 
```

**Reply by PJ5168** [reply]
May I know what's your external I2S DAC?

**Reply by madias123** [reply]
Sorry, its in the // _> PT8211 
Edit, got it:
SetLsbJustified is **not** implemented in the Audio Library for RP2040, must set manually, for a quick/dirty hack, I inserted in line 302, AudioOutputI2S.cpp
```
 #elif defined(ARDUINO_ARCH_RP2040)
    (void)txDAC;
    if (!i2sOn) {
        i2s.setBCLK(bclkPin);
	i2s.setDATA(doutPin);
	i2s.setLSBJFormat(); // Dirty hack for PT8211
        i2s.begin(hertz);
		i2s.setLSBJFormat();
    }
  #endif
  i2sOn = true;
  SetRate(hertz); // Default
  return true;
```

**Reply by PJ5168** [reply]
some more testing on the RP2040,
WAV : 
     48k16bit OK
     96k24bit fail, does not support 24bit
     192k16bit fail, reader does not support 192k
FLAC : 
     can play 96k24b, but in slow motion and breaking up as in old core

**Reply by madias123** [reply]
I tested continuous streaming mp3 (various bitrates) from SD card with mixed results.
First: I run  the same dev board (soldered together no bread board) either  with a VS1053 hardware codex or a PT8211 DAC, just for testing the setup and SD card (connection). So the only difference is the output and decoder (HW/SW)
Conclusio: 
Need ad least 200MHZ (and above)  overclocking and it isn`t stable at all. (Sometimes it breaks the audio after one file, sometimes after 3-5 titles)
It won`t crash the kernel (have a dummy blinking code in background, which works even after breaking audio).
So: Playing short sequences might be fine, but for using it as an (all in one) audio-player it`s better to invest some coins into a VS1053 board or using an ESP32.

PS: Using AudioGeneratorMP3a (helix) only gives a stuttering output.

**Reply by PJ5168** [reply]
Sorry that I don't have PT8211 in hand. The DAC I tested are PCM5102A, MAX98357A, and UDA1334A.
Following FYR:
1. Usually, run various programs on the RP2040 at 200~250Mhz, so far, no issue encountered.
2. The audio clip used for testing is 9'14" in length. No issue to play it repeatedly.
3. For 48K audio, will have some odd effecs if you set the I2S to 44.1K.
4. The default speed of SD is really really slow. You can change it if you run the RP2040 at 125Mhz. What's your SD speed now?
    Or, you have to overclock the peripheral clock if you run the RP2040 at the clock higher than 125Mhz.
   The current test setting is 240Mhz RP2040, 40Mhz SPI SD.
5. Add buffers will help on high bit rate.
2 cents~

**Reply by madias123** [reply]
Hi, 
I think it doesn't matter which DACs, because they are "fire and forget" (Even the PT8211 "forgives" really everything - timeouts - jitter -  - I build synthesizers in the past with STM32 with this DAC)
SD Card speed is set to 40MHZ (Even 15MHZ is enough for MP3 -tested with the VS1053 module)
The dropouts are not music file depended (even the same file sometimes won*t play anymore in the loop)
as I wrote, it should not be a hardware failure , cause I cross testing it with a VS1053 module (even for the DAC I use the SD-card  interface of the VS1053 module).
So I think, there might be a problem within the audio library, sd-lib or the I2s library. So when I*ve  finished my project, I*m going to build  a pico probe for further stepping in.

**Reply by PJ5168** [reply]
Since I2S is a low speed serial interface, if there is a hardware failure, should be able to be captured by logic analyzer.
More, for cycle-to-cycle jitter or long-term jittering, can be verified by a DSO as well.
Do you think if there is any possible power or timing related hardware issues?
Looking forward to seeing your future findings.
BTW, for cost concern, DFPlayer is cheaper than VS1053 module. How's that?

**Comment by madias123** [comment]
I am currently transferring an older project (RFID TAG player (not only) for children)  that I wrote for STM32duino. (STM32F103c**). There is a series of a good 20 devices that I had sold. Unfortunately, the base (STM32F103c** - "Bluepill") is no longer available (only fake chips) and the libraries of the original software are hopelessly outdated. In the meantime I've rewritten most of the software (different graphics drivers, now u8g2) and it's working again on a RP2040.
The DF Player module is useful, but has several (major) disadvantages:
* There is no direct access to the SD card. (I need the SD card for other files (tags, BMPs, settings...)
* The file names and folder structure are/is severely restricted.
* You have no chance to read the ID3 tags (display)
The extrem handy and powerful VS1503 module is on sale for about 9 Euros and I use the SD card slot of it. 
Sadly my website is german only, but here you can see some (early beta) videos of my first player:
https://diroaudio.wordpress.com/anleitung/
(...and the answer to the question: "Why no ESP32?":
Too few pins, too much power consumption and and a restricted policy of not creating the next "internet of shit" - It's not just the technical implementation, it's about starting a creative process together with children without being on the Internet..)
So my first priority is getting my project back to life :) 
There are other ideas with the RP2040 in future (maybe some synths, I have several analog-filter circuits (and boards) left) - for sure with the DACs.

**Reply by PJ5168** [reply]
Roger that! Seems that the VS1503 is the only one can fit your criteria.
Look at the link. The player is really nice with little OLED.
I know how to id if the 328p is fake or not. And I know there are clone F103C.
The bluepills I bought seems not with clone ones. May I know how to id if the F103C is fake or not?
For power/performance, the F4 is better than F1. And F4 is a little little bit better than RP2040@133Mhz.
For ESP32, if you turn off the WiFi and BLE, the power/performance is good as well.
Especially, ESP32 can go to sleep while playing I2S by using the ultra-low power core.
For synths and SDR, Teensy 4 is really a power house. The Dirtywave M8 is really fun!~

**Reply by madias123** [reply]
There are tooooo many MCUs (esp. teensy) out there and I have to less time for it :) I have a couple of F4 devboards in the "drawer of oblivion."
The biggest advantage of the Pi Pico is, that I get one - totally official - for less than 4 Euro in one day. 
BluePill: There are many test codes and websites for it:
https://mecrisp-stellaris-folkdoc.sourceforge.io/bluepill-diagnostics-v1.6.html
https://embeddedtronicsblog.wordpress.com/2018/12/29/fake-stm32-blue-pill-boards/
<7 years ago, you get legit ones with 128kB flash. Now you get mostly bad ones (yes, they blink, but if you step within DMA you *ll see the problems...)

**Reply by PJ5168** [reply]
Thanks for the links. Knew that there are 64K and 128K parts. For 64K part, the hidden 64K is not guaranteed.
Believe, you still can get the bluepill with genuine STM32F103C8T6, it's just a little bit higher than the "fake" one.
You can buy both from following link: https://www.aliexpress.com/item/32792513237.html

**Reply by madias123** [reply]
Every "genius" STM32F103C8T6 has (had?) unofficially  128K - i don't think, ST changed their production.
Even your linked bluepill is a fake for sure, because the genius chips are sold for at least ~3.42 Euros (250 pieces and up) and have hardly been available for the last three years.
https://tinyurl.com/4b55ex7m
I own around a hundred bluepills from various sources and all of them for the last three years have been fakes/knockoffs.
So the bluepill is really dead, but long lives the blackpill! (STM32F401/F411) - you get them for 1.50 Euro.
Interesting part: Even on mouser the STM32F401 is cheaper than the STM32F103: STM32F401CCF6TR Reel  5000:	€ 2,79
At the moment there are no reports of counterfeits of the STM32F4xx, they should all come from an overproduction (or "grey" production of the original)
